keps/sig-storage/3333-reconcile-default-storage-class/README.md

```diff
@@ -0,0 +1,711 @@
+   default SC for some time and second case above applies here too.
+
+3. Defined ordering during cluster installation. Kubernetes cluster installation
+   tools must be currently smart enough to create a default SC before starting
+   anything that may create PVCs that need it. If such a tool supports multiple
+   cloud providers, storage backends and addons that require storage (such an image
+   registry), it may be quite complicated to do the ordering right.
+
+4. If a default storage class does not exist, PVCs with
+   `pvc.spec.storageClassName=nil` are bound to PVs with
+   `pv.spec.storageClassName=""`. This can be confusing to users, the PVC is bound
+   to different PVs or dynamically provisioned, depending on if the default SC
+   exists or not at the time when the PVC is created.
+
+### Goals
+
+* Loosen ordering requirements between PVCs and default SC creation.
+* Improve user experience with `pvc.spec.storageClassName=nil`.
+
+### Non-Goals
+
+* Introduce new API for default SCs.
+
+## Proposal
+
+Right now, the default SC is applied to PVCs in an admission plugin, i.e. only
+during PVC creation. Later on, PV controller will bind PVCs with
+`pvc.spec.storageClassName=nil` to PVs with `pv.spec.storageClassName=""`,
+if such a PV exists.
+
+We want to change the existing `pvc.spec.storageClassName=nil` behavior
+to always require a default SC. It would never bind to a PV with
+`pv.spec.storageClassName=""` and it would be `Pending` until a default SC
+is created. When a default SC is created, the PVC `pvc.spec.storageClassName`
+will be updated with the new default SC name in the PV controller.
+
+This behavior should be simpler from the user perspective.
+
+We plan to re-use the existing `storageclass.kubernetes.io/is-default-class`
+SC annotation.
+
+### User Stories (Optional)
+
+
+#### Story 1
+
+Admin needs to change the default SC from SC1 to SC2
+1. The admin marks the current default SC1 as non-default.
+2. Another user creates PVC requesting a default SC, by leaving
+   `pvc.spec.storageClassName=nil`. The default SC does not exist at
+   this point, therefore the admission plugin leaves the PVC untouched with
+   `pvc.spec.storageClassName=nil`.
+3. The admin marks SC2 as default.
+4. PV controller, when reconciling the PVC, updates
+   `pvc.spec.storageClassName=nil` to the new SC2.
+5. PV controller uses the new SC2 when binding / provisioning the PVC.
+
+#### Story 2
+
+An installation tool wants to install Kubernetes with a CSI driver providing
+a default SC and an application that wants to use it (such as image registry).
+
+1. The installer creates PVC for the image registry first, requesting the
+   default storage class by leaving `pvc.spec.storageClassName=nil`.
+2. The installer creates a default SC.
+3. PV controller, when reconciling the PVC, updates
+   `pvc.spec.storageClassName=nil` to the new default SC.
+4. PV controller uses the new default SC when binding / provisioning the PVC.
+
```

Comment:
Can you add a story where one default SC exists before a PVC is requested. Does this story stay the same before and after the proposed change?

Reply:
Adding.